### PR TITLE
introduces basigx-form-addarcgisrest

### PR DESCRIPTION
**Pull request by jansule**
This introduces the form for adding layers via the ArcGIS Rest API.

Currently, following features/formats are supported:

- import services of type `MapServer`
- import services of type `ImageServer`
- import mentioned services from any ArcGIS Rest folder

Following features are not (yet?) supported:

- import services of type `FeatureServer`
- import single layers of a service

This is basically a copy of `basigx-form-addwms` with adjustments to fit the ArcGIS Rest API. Therefore, the code might benefit from a minor refactoring in a follow-up PR.

In contrast to `basigx-form-addwms`, the OpenLayers layers will only be created when `Ausgewählte Layer hinzufügen` was clicked. This was done to reduce the number of http requests.

Example import from root folder:

compass-arcgis-rest

Example import from sub folder:

compass-arcgis-rest-subfolder

### CODE REVIEW

File: src/util/ArcGISRest.js [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+        /**
+         * Get the root URL of the ArcGISRest service.
+         *
+         * @param {string} url The URL to get the root URL from.
+         * @return {string} The root URL.
+         */
+        getArcGISRestRootUrl: function(url) {
+            if (!BasiGX.util.ArcGISRest.isArcGISRestUrl(url)) {
+                return;
```
Comment (weskamm) [inline src/util/ArcGISRest.js:48]: minor: could throw a warning or console message

File: src/util/ArcGISRest.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+         * @param {string} layerConfig.url The service url.
+         * @param {boolean} useDefaultHeader Whether to use the default
+         * Xhr header.
+         * @return {Ext.Promise} A promise containing the olLayer.
+         */
+        createOlLayerFromArcGISRest: function(layerConfig, useDefaultHeader) {
+            var service = layerConfig.service;
+            var url = layerConfig.url;
+            var serviceUrl = BasiGX.util.ArcGISRest.getArcGISRestRootUrl(url);
+            serviceUrl = [serviceUrl, service.name, service.type].join('/');
```
Comment (weskamm) [inline src/util/ArcGISRest.js:75]: serviceUrl may be undefined here, could use a check

File: src/util/Url.js [hunk 3]
```diff
@@ -97,6 +97,27 @@ Ext.define('BasiGX.util.Url', {
...
+         *
+         * @param {string} url The url.
+         * @param {string} queryKey The key of the query parameter.
+         * @param {string} queryValue The value of the query parameter.
+         * @return {string} The URL with the new query parameter.
+         */
+        setQueryParam: function(url, queryKey, queryValue) {
+            var urlObj = new URL(url);
+            var searchParams = urlObj.searchParams;
+            searchParams.set(queryKey, queryValue);
```
Comment (weskamm) [inline src/util/Url.js:117]: i think this setter is case sensitive, so please check if you want to handle that or if its ok

File: src/view/button/AddArcGISRest.js [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+ */
+Ext.define('BasiGX.view.button.AddArcGISRest', {
+    extend: 'BasiGX.view.button.Base',
+    xtype: 'basigx-button-addarcgisrest',
+
+    requires: [
+        'Ext.window.Window',
+        'Ext.app.ViewModel',
+        'BasiGX.view.form.AddArcGISRest',
+        'BasiGX.util.Animate'
```
Comment (weskamm) [inline src/view/button/AddArcGISRest.js:32]: nice!

File: src/view/form/AddArcGISRest.js [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+     * ArcGISRest folder. This method will examine the answer
+     * and eventually set up a fieldset for all the services that
+     * we have found in the server's answer.
+     *
+     * @param {XMLHttpRequest} response The response of the request.
+     */
+    onGetServicesSuccess: function(response) {
+        try {
+            var responseJson = JSON.parse(response.responseText);
+            var hasServicesKey = Object.prototype.hasOwnProperty.call(
```
Comment (weskamm) [inline src/view/form/AddArcGISRest.js:338]: minor: could have used `Object.keys().includes` but should lead to the same result...

**Review by weskamm — APPROVED** [review]
Nice work! I just added some minor comments.

File: src/util/Url.js [hunk 6]
```diff
@@ -97,6 +97,27 @@ Ext.define('BasiGX.util.Url', {
          */
         isUrl: function (str) {
             return str.startsWith('http://') || str.startsWith('https://');
+        },
+
+        /**
+         * Set a query parameter.
```
Comment (chrismayer) [inline src/util/Url.js:103]: `Sets` instead of `Set`?

File: src/view/button/AddArcGISRest.js [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    ],
+
+    /**
+     *
+     */
+    viewModel: {
+        data: {
+            tooltip: 'ArcGIS Rest Layer hinzufügen',
```
Comment (chrismayer) [inline src/view/button/AddArcGISRest.js:40]: Is `Rest` the common writing or should it be `REST`?

**Review by chrismayer — APPROVED** [review]
Very nice @jansule. Thanks for this. Only minor wording stuff from my side.

File: src/util/ArcGISRest.js (hunk 1 above)
Comment (jansule) [reply src/util/ArcGISRest.js:48]: solved in https://github.com/terrestris/BasiGX/pull/691/commits/a6cdee5a6ed6a5a5d61f245a092086eb9779c626

File: src/util/Url.js (hunk 3 above)
Comment (jansule) [reply src/util/Url.js:117]: In our case, we should consider case-sensitiviy. So I will keep it as is.

File: src/util/Url.js (hunk 6 above)
Comment (jansule) [reply src/util/Url.js]: solved in https://github.com/terrestris/BasiGX/pull/691/commits/a6cdee5a6ed6a5a5d61f245a092086eb9779c626

File: src/view/button/AddArcGISRest.js (hunk 7 above)
Comment (jansule) [reply src/view/button/AddArcGISRest.js]: adjusted in https://github.com/terrestris/BasiGX/pull/691/commits/a6cdee5a6ed6a5a5d61f245a092086eb9779c626

**Pull request merged**